### ISSUE TITLE
Refactor configuration layout

### ISSUE THREAD

**Issue by aeneasr** [ory/kratos#432]
**Is your feature request related to a problem? Please describe.**

The current configuration layout is very verbose. It also requires extra attention to the docs which in turn leads to bugs that people find hard to figure out (e.g. using redirect after registration which disables registration effectively).

It's not very helpful that, if e.g. email verification is enabled, one needs to set up the appropriate after hooks for all the flows. Same thing applies to sign up with oidc and the session handler. This should be handled centrally, by a central switch.

Additionally, we need a secret to hash and encrypt things. Currently, only a session secret can be set. Using one secret for the whole system though is preferred.

The `urls` section is also kind of confusing, it would be better to split this into each system.

**Describe the solution you'd like**

We should split the config definitions into:
 
- `selfservice.flows`: Core user interactions ("selfservice flows") that must always be there for the basic system to work - login, logout, registration, settings. Additional, optional user interactions such as account `verficiation` and account `activation` and `mfa` can be enabled and configured if needed.
- `selfservice.strategies`: Strategies implement flows such as `password` and `oidc`.

Configuring the self-service potion looked like this before. Using a very verbose config to make the point and annotations to hint at problems:

```yaml
selfservice:
  strategies:
    password:
      enabled: true
    oidc:
      enabled: true
      config: # <- why is this here? we can make providers a direct child of oidc.
         providers:
           # - id: github ...
           - id: google # this is `<provider-id>`
             provider: google
             client_id: ....
             client_secret: ...
             mapper_url: file:///etc/config/kratos/oidc.google.jsonnet
             scope:
               - email

  settings:
    privileged_session_max_age: 1m
    # All of these hooks don't really make sense to have configureable. Instead, if verification
    # is on, this is done automatically.
    after:
      profile:
        hooks:
          - hook: verify # <- required or verification will not work when updating email addresses
          - hook: recovery # <- required or recovery will not work when updating email addresses

  verify:
    return_to: http://127.0.0.1:4455/

  recovery:
    return_to: http://127.0.0.1:4455/

  logout:
    redirect_to: http://127.0.0.1:4455/auth/login

  login:
    request_lifespan: 10m
    # revoke_active_session is more or less the only useful hook here at the moment
    after:
      password:
        hooks:
          - hook: revoke_active_sessions
      oidc:
        hooks:
          - hook: revoke_active_sessions


  registration:
    request_lifespan: 10m
    # Here too we need to define a lot of stuff for recovery/verification to work. If we don't do this
    # those features might / will break.
    after:
      oidc:
        hooks:
          - hook: session # <- without this the user would have to do the same oidc flow again just to be logged in now.
          - hook: verify # <- without this verification would break - and we can't use email_verified from the openid provider.
          - hook: recovery # <- without this recovery is disabled / would not work
      password:
        hooks:
          - hook: session # <- only useful hook here
          - hook: verify # <- without this verification would break.
          - hook: recovery # <- without this recovery is disabled / would not work

# This section is pretty senseless. Instead, these urls should be defined for the flows. This makes it also much easier to leave them empty if a feature (e.g. recovery) is disabled.
urls:
  login_ui: http://127.0.0.1:4455/auth/login
  registration_ui: http://127.0.0.1:4455/auth/registration
  error_ui: http://127.0.0.1:4455/error
  settings_ui: http://127.0.0.1:4455/settings
  verify_ui: http://127.0.0.1:4455/verify # <- required even if you don't even use verification
  recovery_ui: http://127.0.0.1:4455/recovery # <- required even if you don't even use recovery
  mfa_ui: http://127.0.0.1:4455/ # <- required even if you don't even use verification

  # This really should be named `base_url` because that's what it is.
  self:
    public: http://127.0.0.1:4455/.ory/kratos/public/ # <- required even in cases where this could be auto-detected from the HTTP request
    admin: http://kratos:4434/ # <- required even in cases where this could be auto-detected from the HTTP request
  default_return_to: http://127.0.0.1:4455/
  whitelisted_return_to_urls:
    - http://127.0.0.1:4455

courier:
  smtp:
    connection_uri: smtps://test:test@mailslurper:1025/?skip_ssl_verify=true # <- required even if you don't even use verification / recovery
```

**The new config would look like this:**

```yaml
selfservice:
  flow:
    settings:
      ui_url: http://127.0.0.1:4455/settings # previously urls.settings_ui
      # privileged_session_max_age: 1m

      # If the client is a browser, redirect here after a successful update:
      # return_browser_to: http://127.0.0.1:4455/dashboard

      # there will still be an after hook here but since there are currently no hooks that make sense here the section
      # is omitted. `verification` and `activation` are now applied automatically when enabled so no need to define them
      # here.

    logout:
      # If the client is a browser, redirect here after a successful logout:
      # return_browser_to: http://127.0.0.1:4455/auth/login

    login:
      ui_url: http://127.0.0.1:4455/auth/login # previously urls.login_ui
      # request_lifespan: 10m

      # If the client is a browser, redirect here after a successful login:
      # return_browser_to: http://127.0.0.1:4455/dashboard

      # These two remain the same:
      #
      # before: ...
      # after: ...

    registration:
      ui_url: http://127.0.0.1:4455/auth/registration # previously urls.registration_ui
      # request_lifespan: 10m

      # If the client is a browser, redirect here after a successful registration:
      # return_browser_to: http://127.0.0.1:4455/dashboard

      # The before section remains the same:
      # before

      # The after section stays very similar to the previous version with the exception that verification/recovery
      # hooks no longer exist and that oidc now automatically issues a session.
      after:
        password:
          hooks:
            # This still ogs the user in after successful registration using password:
            - hook: session
        oidc:
          hooks:
          # The session hook is now always active for oidc post-registration so no need to define it.

          # You might still have future hooks here:
          # - hook: some_hook

    verification:
      enabled: true # set false (default) to disable email verification. If enabled, each identity MUST have at
                    # least one trait that is an email address and that can be verified.
      ui_url: http://127.0.0.1:4455/verify # previously urls.verify_ui - required if enabled

      # If the client is a browser, redirect here after a successful verification:
      # return_browser_to: http://127.0.0.1:4455/auth/login

      # In a future version we might want to add more things such as only allowing login for verified accounts:
      # login_only_verified: true

    # This flow is new!
    recovery:
      enabled: true # disables account recovery - makes sense when your user's don't have an email.
                    # If enabled, each identity MUST have at
                    # least one trait that is an email address and that can be verified.
      ui_url: http://127.0.0.1:4455/recover # required if enabled

      # In a future version we might add additional things such as security questions:
      # security_questions:
      #   - id: foobar
      #     label: foobarbaz

      # If the client is a browser, redirect here after a successful verification:
      # return_browser_to: http://127.0.0.1:4455/auth/login

  strategies:
    password:
      # The password strategy remained the same.
      enabled: true

    oidc:
      enabled: true

      # config: <- this key is has been removed! providers is now a direct child of `oidc`!
      providers:
        # - id: github ...
        - id: google # this is `<provider-id>`
          provider: google
          client_id: ....
          client_secret: ...
          mapper_url: file:///etc/config/kratos/oidc.google.jsonnet
          scope:
            - email

secrets:
  default: # this was previously `cookie`
    - PLEASE-CHANGE-ME-I-AM-VERY-INSECURE

# This section was previously this:
#   urls:
#     default_return_to: http://127.0.0.1:4455/
#     whitelisted_return_to_urls:
#       - http://127.0.0.1:4455
return_to:
  default: http://127.0.0.1:4455/
  whitelisted:
    - http://127.0.0.1:4455

serve:
  public:
    base_url: http://127.0.0.1:4455/.ory/kratos/public/ # this was previously urls.self.public
  admin:
    base_url: http://kratos:4434/ # this was previously urls.self.admin
```

While there is still lots of stuff configurable, the overall config is much leaner and leaves less room for error as you can see in the last example which is the previous config without all the comments:

```yaml
selfservice:
  flow:
    settings:
      ui_url: http://127.0.0.1:4455/settings
      privileged_session_max_age: 1m
      return_browser_to: http://127.0.0.1:4455/dashboard

    logout:
      return_browser_to: http://127.0.0.1:4455/auth/login

    login:
      ui_url: http://127.0.0.1:4455/auth/login

    registration:
      ui_url: http://127.0.0.1:4455/auth/registration
      request_lifespan: 10m
      return_browser_to: http://127.0.0.1:4455/dashboard
      after:
        password:
          hooks:
            - hook: session

    verification:
      enabled: true
      ui_url: http://127.0.0.1:4455/verify

    recovery:
      enabled: true
      ui_url: http://127.0.0.1:4455/recover
      return_browser_to: http://127.0.0.1:4455/auth/login

  strategies:
    password:
      enabled: true

    oidc:
      enabled: true
      providers:
        # - id: github ...
        - id: google # this is `<provider-id>`
          provider: google
          client_id: ....
          client_secret: ...
          mapper_url: file:///etc/config/kratos/oidc.google.jsonnet
          scope:
            - email

secrets:
  default:
    - PLEASE-CHANGE-ME-I-AM-VERY-INSECURE

return_to:
  default: http://127.0.0.1:4455/
  whitelisted:
    - http://127.0.0.1:4455

serve:
  public:
    base_url: http://127.0.0.1:4455/.ory/kratos/public/
  admin:
    base_url: http://kratos:4434/
```

**Describe alternatives you've considered**

Happy to discuss alternatives!

**Comment by koenmtb1** [comment]
I think a default return_to URL for each flow is pretty useful instead of a global one.

**Comment by aeneasr** [comment]
That's currently also the case! See here: https://www.ory.sh/kratos/docs/concepts/selfservice-flow-completion

**Comment by koenmtb1** [comment]
Oh yeah I see, but what is the global "return_to -> default" for then?

**Comment by aeneasr** [comment]
Is that not clear from this section: https://www.ory.sh/kratos/docs/concepts/selfservice-flow-completion/#redirection ? If so, how could this be improved to make this easier to understand?

**Comment by aeneasr** [comment]
**Please ignore this!**

---

## Related issue

#432

## Proposed changes

### Admin and Public Base URLs

```diff
- urls:
-   self:
-     public: https://kratos.my-website.com/
-     admin: https://admin.kratos.cluster.localnet/
+ serve:
+   public:
+     base_url: https://kratos.my-website.com/
+   admin:
+     base_url: https://admin.kratos.cluster.localnet/
```

### Session Secrets


```diff
- secrets:
-   session:
-     - secret-to-encrypt-session-cookies
-     - old-session-cookie-secret-that-has-been-rotated
+ secrets:
+  default:
+    # This secret is used as default and will also be used for encrypting e.g. cookies when a dedicated cookie secret (as shown below) is not defined.
+    - default-secret-to-encrypt-stuff
+   cookie:
+     - secret-to-encrypt-session-cookies
+     - old-session-cookie-secret-that-has-been-rotated
```

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

## Further comments
